src/charm.py

```diff
@@ -1086,71 +1066,46 @@ def _create_raft_snapshot(self) -> Optional[IO[bytes]]:
             return None
-        root_token, unseal_keys = self._get_initialization_secret()
-        vault.authenticate(Token(root_token))
-        vault.unseal(unseal_keys=unseal_keys)
         response = vault.create_snapshot()
         return response.raw
 
-    def _restore_vault(
-        self, snapshot: StreamingBody, restore_unseal_keys: List[str], restore_root_token: str
-    ) -> bool:
+    def _restore_vault(self, snapshot: StreamingBody) -> bool:
         """Restore vault using a raft snapshot.
 
-        Updates the initialization secret.
-        Upon successful secret update, it will restore the raft snapshot.
-        Upon successful restore, it will unseal Vault and set root token.
-
         Args:
             snapshot: Snapshot to be restored as a StreamingBody from the S3 storage.
-            restore_unseal_keys: List of unseal keys used at the time of the backup.
-            restore_root_token: Root token used at the time of the backup.
 
         Returns:
             bool: True if the restore was successful, False otherwise.
         """
-        if not (vault := self._get_initialized_vault_client()):
+        if not (vault := self._get_active_vault_client()):
             logger.error("Failed to get Vault client, cannot restore snapshot.")
             return False
-        (
-            current_root_token,
-            current_unseal_keys,
-        ) = self._get_initialization_secret()
-        vault.authenticate(Token(current_root_token))
-        vault.unseal(unseal_keys=current_unseal_keys)
-
-        self._set_initialization_secret(
-            root_token=restore_root_token, unseal_keys=restore_unseal_keys
-        )
+
         try:
             # hvac vault client expects bytes or a file-like object to restore the snapshot
             # StreamingBody implements the read() method
             # so it can be used as a file-like object in this context
             response = vault.restore_snapshot(snapshot)  # type: ignore[arg-type]
         except Exception as e:
             # If restore fails for any reason, we reset the initialization secret
-            logger.error("Failed to restore snapshot: %s", e)
-            self._set_initialization_secret(
-                root_token=current_root_token, unseal_keys=current_unseal_keys
-            )
+            logger.error("Failed to restore snapshote: %s", e)
             return False
         if not 200 <= response.status_code < 300:
-            logger.error("Failed to restore snapshot: %s", response.json())
-            self._set_initialization_secret(
-                root_token=current_root_token, unseal_keys=current_unseal_keys
-            )
+            logger.error("Failed to restore snapshotes: %s", response.json())
             return False
-        vault.authenticate(Token(restore_root_token))
-        vault.unseal(unseal_keys=restore_unseal_keys)
+
+        # Check if the approle can still login
+        # If not, remove secret
```

Comment:
Why those comments here?

Reply:
this is a discussion I'd like to have with the team, I should have marked it as such. Instead, I'll remove it and create a ticket to address it later (this includes the above change with force_restore_snapshot). The gist of it is the charm will not be able to warn the user properly right now if they restore a vault that doesn't have the charm policy and role that was created when they first started vault. The options are either we handle that case, or we stop forcing snapshot restores.

Reply:
Just going to leave my comments from a private conversation I had with Kayra:

I think it's reasonable for us to go into a blocked status when this happens. If they happen to restore a different Vault, or a Vault without the approle policy, we can force them to re-run the `authorize-charm` action which will either create or update the policy. They will need a token (likely the root token) to do this.

Reply:
The functionality in the comments is here now.